### ISSUE TITLE
NCuda memory error when training with QueryCrossEntropy loss

### ISSUE THREAD

**Issue by tigromol** [catboost/catboost#2372]
Problem:
NCuda memory error when using QueryCrossEntropy loss 
and no error when using PairlogitPairwise loss on the same dataset
catboost version: 1.0.6, 1.0.4
Operating System: 
"Debian GNU/Linux 10 (buster)"
CPU:
64 x Intel(R) Xeon(R) Gold 6226R CPU @ 2.90GHz
GPU:
4x nvidia a100

NVIDIA-SMI 510.47.03    
Driver Version: 510.47.03   
CUDA Version: 11.6

**Comment by tigromol** [comment]
train dataset shape:
100kk - rows 
53 features
50 numeric + 3 categorial

**Comment by tigromol** [comment]
Параметры для трейна:
"train_params": {
        "boosting_type": "Plain",
        "max_ctr_complexity": 1,
        "objective": "QueryCrossEntropy",
        "eval_metric": "NDCG:top=30",
        "early_stopping_rounds": 200,
        "task_type": "GPU",
        "gpu_ram_part": 0.8,
        "num_trees": 2000,
        "metric_period": 25,
        "learning_rate": 0.2,
        "devices": [
            0,
            1,
            2,
            3
        ]
    },

**Comment by tigromol** [comment]
Tried with catboost:1.2
Got same error:

Application terminated with error: ??+0 (0x7F679CE4F0C0)
??+0 (0x7F679CE4BCCE)
??+0 (0x7F679CE4B9F6)
??+0 (0x7F679BAFC185)
??+0 (0x7F679B9E2F50)
??+0 (0x7F679CE4C6B7)
??+0 (0x7F679CE4DAD8)
??+0 (0x7F679BAFCB97)
??+0 (0x7F679BAFCA47)
??+0 (0x7F685C4A6FA3)
clone+63 (0x7F685C24B06F)

(NCudaLib::TOutOfMemoryError) /src/catboost/catboost/cuda/cuda_lib/memory_pool/stack_like_memory_pool.h:303: Error: Out of memory. Requested 11830.02954 MB; Free 9205.699804 MB
uncaught exception:
    address -> 0x2827b930410
    what() -> "/src/catboost/catboost/cuda/cuda_lib/memory_pool/stack_like_memory_pool.h:303: Error: Out of memory. Requested 11830.02954 MB; Free 9205.699804 MB"
    type -> NCudaLib::TOutOfMemoryError

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@tigromol 

> train dataset shape: 100kk - rows 53 features 50 numeric + 3 categorial

unlike PairLogitPairwise, QueryCrossEntropy requires gpu memory to store, for each group, the upper triangle of a square matrix

most likely, this is the root-cause
